app/main/helpers/frameworks/validation.py

```diff
@@ -0,0 +1,178 @@
+from functools import reduce
+from operator import add
+import re
+import six
+from werkzeug.datastructures import ImmutableOrderedMultiDict
+
+EMAIL_REGEX = r'^[^@^\s]+@[^@^\.^\s]+(\.[^@^\.^\s]+)+$'
+
+G7_CHARACTER_LIMIT = 5000
+
+G7_OPTIONAL_FIELDS = set([
+    "SQ1-1p-i", "SQ1-1p-ii", "SQ1-1p-iii", "SQ1-1p-iv",
+    "SQ1-1q-i", "SQ1-1q-ii", "SQ1-1q-iii", "SQ1-1q-iv", "SQ1-1cii", "SQ1-1i-ii",
+    "SQ1-1j-i", "SQ1-1j-ii", "SQ4-1c", "SQ3-1k", "SQ1-1i-i"
+])
+
+G7_EMAIL_VALIDATION_FIELDS = set(['SQ1-1o', 'SQ1-2b'])
+
+validators = {}
+
+
+def get_validator(framework):
+    """
+    Retrieves a validator by name, otherwise returns EmptyValidator.
+    """
+    if framework is not None:
+        framework_name = framework.get('slug', None)
+        return validators.get(framework_name, EmptyValidator)
+    else:
+        return EmptyValidator
+
+
+class FrameworkValidator(object):
+    """
+    Base class for validators. Don't call directly.
+
+    Subclass and override the errors method to add whatever rules
+    you wish to have for a particular framework.
+    """
+    email_validation_fields = []
+    character_limit = None
+
+    def __init__(self, content, answers):
+        self.content = content
+        self.answers = answers
+
+    def get_error_messages_for_page(self, section):
+        all_errors = self.get_error_messages()
+        page_ids = section.get_question_ids()
+        page_errors = ImmutableOrderedMultiDict(filter(lambda err: err[0] in page_ids, all_errors))
+        return page_errors
+
+    def get_error_messages(self):
+        raw_errors_map = self.errors()
+        errors_map = list()
+        for question_number, question_id in enumerate(self.all_fields()):
+            if question_id in raw_errors_map:
+                validation_message = self.get_error_message(question_id, raw_errors_map[question_id])
+                errors_map.append((question_id, {
+                    'input_name': question_id,
+                    'question': "Question {}".format(question_number + 1),
+                    'message': validation_message,
+                }))
+
+        return errors_map
+
+    def get_error_message(self, question_id, message_key):
+        for validation in self.content.get_question(question_id).get('validations', []):
+            if validation['name'] == message_key:
+                return validation['message']
+        default_messages = {
+            'answer_required': 'Answer required'
+        }
+        return default_messages.get(
+            message_key, 'There was a problem with the answer to this question')
+
+    def errors(self):
+        return {}
```

Comment:
Since it's meant to be overriden, `errors` should `raise NotImplementedError` to avoid the possibility of accidentally calling a class that doesn't actually do any validation.